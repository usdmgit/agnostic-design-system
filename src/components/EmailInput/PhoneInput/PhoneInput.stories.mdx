# Phone Input Component

## Type: Simple

## Phone Input Props

|        Prop                 |           Type            |                                                                                                                                      Description                                                                                                                                       |        Default         | Mandatory |
| :-------------------------: | :-----------------------: | :------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: | :--------------------: | :-------: |
|        autoFormat           |          boolean          |                                                                                                              Specifies whether or not phone formatting should be enabled.                                                                                                              |          true          |   false   |
|         country             |          string           |                                                                                                                                 Initial country code.                                                                                                                                  |           -            |   false   |
|         defaultMask         |          string           |                                                                                                              Default mask to be used in case no other mask is specified (Our default is the US mask).                                                                                  |    '(...) ...-....'    |   false   |
|         disabled            |          boolean          |                                                                                                              Specifies whether or not the Phone Input should be disabled.                                                                                                              |         false          |   false   |
|      disableDropdown        |          boolean          |                                                                                                          Specifies whether or not the Countries Dropdown should be disabled.                                                                                                           |         false          |   false   |
|        enableSearch         |          boolean          |                                                                                                                       Enables search in the Countries Dropdown.                                                                                                                        |         false          |   false   |
|      excludeCountries       |         string[]          |                                                                                                                             Country codes to be excluded.                                                                                                                              |           -            |   false   |
|         hideLabel           |          boolean          |                                                                                                                  Specifies whether or not the label should be hidden.                                                                                                                  |         false          |   false   |
|     variablesClassName      |          string           |                                                                                                                  Receives the class for the actual input to customize its style.                                                                                                       |           -            |   false   |
|           label             | string \| React.ReactNode | Can be used as a [string](https://design-system.codelitt.dev/?path=/story/components-input--default-with-label-text) and allow the engineer to build a [custom label](https://design-system.codelitt.dev/?path=/story/components-input--with-custom-label) and pass it as a parameter. |           -            |   false   |
|        localNumber          |          boolean          |                          Specifies whether the number is local and should default to the country you added or if the API should try to guess the Country you are adding a number for.                                                                                                  |           -            |   false   |
|           masks             |    Record<string,string>  |                                                                                                  Defines a map of masks and its associated country, so you can overwrite the default mask if necessary.                                                                                |           -            |   false   |
|          onChange           |  (value: string) => void  |                                                                                                     Calls a function whenever a change in an event of the Phone Input is detected.                                                                                                     |           -            |   false   |
|       onlyCountries         |         string[]          |                                                                                                                             Country codes to be included.                                                                                                                              |           -            |   false   |
|        placeholder          |          string           |                                                                                                                             Defines a custom placeholder.                                                                                                                              | The default phone mask |   false   |
|     preferredCountries      |         string[]          |                                                                                                                            Country codes to be at the top.                                                                                                                             |           -            |   false   |
|            size             |           Size            |                                                                                                             Size of the component. The change is reflected on the height.                                                                                                              |         large          |   true    |
|           value             |          string           |                                                                                                                          The value of the Phone Input field.                                                                                                                           |           -            |   false   |

## Getting Started

```javascript
import React, { useState } from 'react';
import { PhoneInput } from '@codelittinc/agnostic-design-system';

const SimplePhoneInput = () => {
  const [value, setValue] = useState('');

  return <PhoneInput onChange={phone => setValue(phone)} label='Phone Input' value={value} />;
};

export default SimplePhoneInput;
```

## CSS Variables Available

Although it is not a decorator component, we use all the CSS variables of the [Input](https://design-system.codelitt.dev/?path=/docs/components-input--default) component. Please refer to the previous for more details.
If you need to specify or customize the input with your own class, the component will receive an variablesClassName, which will allow you to overwrite the styles of our default input compontent.

PhoneInput variables are listed in the table below.

### Variables of the PhoneInput component

|      Property      |     Attribute      | State |
| :----------------: | :----------------: | :---: |
|      dropdown      |  background-color  |       |
|      dropdown      |       border       |       |
|      dropdown      |   border-radius    |       |
|      dropdown      |       height       |       |
|      dropdown      |       margin       |       |
|      dropdown      |       width        |       |
