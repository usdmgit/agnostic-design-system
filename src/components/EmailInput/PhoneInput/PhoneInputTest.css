.phone-input {
  --input-color: red;
  --input-prepend-padding-left: 58px;
  --phone-input-dropdown-background-color: var(--structure-color-white);
  --phone-input-dropdown-border-radius: 3px;
  --phone-input-dropdown-margin: 4px;
}
