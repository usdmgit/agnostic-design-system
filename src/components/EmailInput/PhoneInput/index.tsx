import React from 'react';
import ExternalPhoneInput from 'react-phone-input-2';
import classNames from 'classnames';
import styles from './PhoneInput.css';
import inputStyles from '@/components/Input/Input.css';

/* eslint-disable */
import { Size } from '../Input';
/* eslint-enable */
interface Props {
  autoFormat?: boolean;
  country?: string;
  defaultMask?: string;
  disabled?: boolean;
  disableDropdown?: boolean;
  enableSearch?: boolean;
  excludeCountries?: string[];
  hideLabel?: boolean;
  label?: string | React.ReactNode;
  localNumber?: boolean;
  masks?: Record<string, string>;
  onChange: (value: string) => void;
  onlyCountries?: string[];
  placeholder?: string;
  preferredCountries?: string[];
  size: Size;
  value?: string;
  variablesClassName?: string;
}

const PhoneInput = (props: Props) => {
  const {
    autoFormat,
    country,
    defaultMask = '(...) ...-....',
    disabled,
    disableDropdown,
    enableSearch,
    excludeCountries,
    hideLabel,
    label,
    localNumber,
    masks,
    onChange,
    onlyCountries,
    placeholder,
    preferredCountries,
    size,
    value,
    variablesClassName
  } = props;

  const derivedDefaultMask = (masks && country && masks[country]) || defaultMask;
  // We need to force a re-render when masks change, otherwise they wont apply correctly
  const key = `${country}-${derivedDefaultMask}-${localNumber}`;
  const sizeClass = `input--${size}`;

  const getLabel = () => {
    return React.isValidElement(label)
      ? label
      : label && (
          <label
            className={classNames(
              inputStyles['input--label'],
              hideLabel && inputStyles['input--label-hidden']
            )}
          >
            {label}
          </label>
        );
  };

  return (
    <>
      {getLabel()}
      <ExternalPhoneInput
        alwaysDefaultMask={localNumber}
        autoFormat={autoFormat}
        buttonStyle={localNumber ? { display: 'none' } : undefined}
        defaultMask={derivedDefaultMask}
        disabled={disabled}
        disableCountryCode={localNumber}
        disableDropdown={disableDropdown}
        enableSearch={enableSearch}
        inputClass={classNames(
          inputStyles.input,
          inputStyles[sizeClass],
          !localNumber && inputStyles['input-with-prepend'],
          variablesClassName
        )}
        containerClass={classNames(styles['react-tel-input'], variablesClassName)}
        excludeCountries={excludeCountries}
        key={key}
        onlyCountries={onlyCountries}
        placeholder={placeholder}
        preferredCountries={preferredCountries}
        country={country}
        value={value}
        onChange={onChange}
      />
    </>
  );
};

export default PhoneInput;

PhoneInput.defaultProps = {
  size: 'large',
  onChange: e => e
};
