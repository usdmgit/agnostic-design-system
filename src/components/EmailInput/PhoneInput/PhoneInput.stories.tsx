import React, { useState } from 'react';
import PhoneInput from '@/components/PhoneInput';
import { Meta } from '@storybook/react';
import mdx from './PhoneInput.stories.mdx';
import classNames from 'classnames';

import styles from './PhoneInputTest.css';

export default {
  title: 'Components/Phone Input',
  component: PhoneInput,
  parameters: {
    docs: {
      page: mdx
    }
  }
} as Meta;

const Template = args => {
  const [value, setValue] = useState('');
  return <PhoneInput {...args} value={value} onChange={e => setValue(e)} />;
};

export const Default = Template.bind({});
Default.args = {
  country: 'us',
  label: 'Default phone input'
};

export const LocalPhoneNumber = Template.bind({});
LocalPhoneNumber.args = {
  country: 'us',
  localNumber: true,
  placeholder: '(702) 123-4567',
  label: 'Local phone number input',
  defaultMask: '(...) ...-....'
};

export const CustomPhoneNumber = Template.bind({});
CustomPhoneNumber.args = {
  country: 'br',
  localNumber: true,
  placeholder: '(99) 99999-9999',
  label: 'Custom Mask phone input',
  masks: { br: '(..) .....-....' }
};

export const VariablesClassName = Template.bind({});
VariablesClassName.args = {
  country: 'us',
  label: 'Default phone input',
  variablesClassName: classNames(styles['phone-input'])
};
