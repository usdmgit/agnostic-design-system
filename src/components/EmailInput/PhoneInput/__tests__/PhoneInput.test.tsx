import React from 'react';
import PhoneInput from '@/components/PhoneInput';
import { mount } from 'enzyme';
import { render, fireEvent } from '@testing-library/react';

describe('PhoneInput', () => {
  it('renders correctly', () => {
    const wrapper = mount(<PhoneInput />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('Local number', () => {
    it('defaults to US mask', () => {
      const utils = render(<PhoneInput placeholder='(999) 999-9999' localNumber />);

      const input = utils.getByPlaceholderText('(999) 999-9999') as HTMLInputElement;

      fireEvent.change(input, { target: { value: '2376324232' } });
      const expectedText = '(237) 632-4232';
      expect(input.value).toBe(expectedText);

      fireEvent.change(input, { target: { value: '5606786876' } });
      const expectedText2 = '(560) 678-6876';
      expect(input.value).toBe(expectedText2);

      fireEvent.change(input, { target: { value: '7698076867' } });
      const expectedText3 = '(769) 807-6867';
      expect(input.value).toBe(expectedText3);
    });

    it('can receive different masks', () => {
      const utils = render(
        <PhoneInput
          placeholder='(99) 99999-9999'
          localNumber
          country='br'
          masks={{ br: '(..) .....-....' }}
        />
      );

      const input = utils.getByPlaceholderText('(99) 99999-9999') as HTMLInputElement;

      fireEvent.change(input, { target: { value: '65992168563' } });
      const expectedText = '(65) 99216-8563';
      expect(input.value).toBe(expectedText);

      fireEvent.change(input, { target: { value: '23423423424' } });
      const expectedText2 = '(23) 42342-3424';
      expect(input.value).toBe(expectedText2);

      fireEvent.change(input, { target: { value: '51132357875' } });
      const expectedText3 = '(51) 13235-7875';
      expect(input.value).toBe(expectedText3);
    });
  });
});
