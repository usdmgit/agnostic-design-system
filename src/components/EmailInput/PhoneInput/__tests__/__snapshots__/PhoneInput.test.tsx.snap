// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhoneInput renders correctly 1`] = `
<PhoneInput
  onChange={[Function]}
  size="large"
>
  <t
    alwaysDefaultMask={false}
    areaCodes={null}
    autoFormat={true}
    autocompleteSearch={false}
    buttonClass=""
    buttonStyle={Object {}}
    className=""
    containerClass="react-tel-input"
    containerStyle={Object {}}
    copyNumbersOnly={true}
    country=""
    countryCodeEditable={true}
    defaultErrorMessage=""
    defaultMask="(...) ...-...."
    disableCountryCode={false}
    disableCountryGuess={false}
    disableDropdown={false}
    disableInitialCountryGuess={false}
    disableSearchIcon={false}
    disabled={false}
    dropdownClass=""
    dropdownStyle={Object {}}
    enableAreaCodeStretch={false}
    enableAreaCodes={false}
    enableClickOutside={true}
    enableLongNumbers={false}
    enableSearch={false}
    enableTerritories={false}
    excludeCountries={Array []}
    flagsImagePath="./flags.png"
    inputClass="input input--large input-with-prepend"
    inputProps={Object {}}
    inputStyle={Object {}}
    isValid={true}
    jumpCursorToEnd={true}
    key="undefined-(...) ...-....-undefined"
    keys={
      Object {
        "A": 65,
        "DOWN": 40,
        "ENTER": 13,
        "ESC": 27,
        "LEFT": 37,
        "PLUS": 43,
        "RIGHT": 39,
        "SPACE": 32,
        "TAB": 9,
        "UP": 38,
        "Z": 90,
      }
    }
    localization={Object {}}
    masks={null}
    onChange={[Function]}
    onEnterKeyPress={null}
    onlyCountries={Array []}
    placeholder="1 (702) 123-4567"
    preferredCountries={Array []}
    prefix="+"
    preserveOrder={Array []}
    priority={null}
    regions=""
    renderStringAsFlag=""
    searchClass=""
    searchNotFound="No entries to show"
    searchPlaceholder="search"
    searchStyle={Object {}}
    showDropdown={false}
    specialLabel="Phone"
    value=""
  >
    <div
      className="react-tel-input "
      onKeyDown={[Function]}
      style={Object {}}
    >
      <div
        className="special-label"
      >
        Phone
      </div>
      <input
        className="form-control input input--large input-with-prepend"
        disabled={false}
        onBlur={[Function]}
        onChange={[Function]}
        onClick={[Function]}
        onCopy={[Function]}
        onDoubleClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        placeholder="1 (702) 123-4567"
        style={Object {}}
        type="tel"
        value=""
      />
      <div
        className="flag-dropdown "
        style={Object {}}
      >
        <div
          aria-haspopup="listbox"
          className="selected-flag"
          onClick={[Function]}
          role="button"
          tabIndex="0"
          title=""
        >
          <div
            className="flag 0"
          >
            <div
              className="arrow"
            />
          </div>
        </div>
      </div>
    </div>
  </t>
</PhoneInput>
`;
